Add tags landing

lickem22 opened the pull request:
Reviewer:@sidravi
Estimate: 1h

---

## Ticket

Fixes: [AAQ-549](https://idinsight.atlassian.net/browse/AAQ-549)

## Description

### Goal
Th Goal of this PR is to allow filtering of contents using content tags and allow to add new tags within the tags bar. 
### Changes
Changes on the backend: No changes
Changes on the front-end:
- Added the Tags Bar on the landing page
- Added filtering feature using the Tag bar
- Updated the Tags Bar on the edit page to add a new tag when the text in the search doesn't return anything. Just like this:
<img width="667" alt="Screenshot 2024-05-30 at 17 45 47" src="https://github.com/IDinsight/aaq-core/assets/44327443/947eb12f-a6ee-46f3-8f71-c069876234fc">
<img width="656" alt="Screenshot 2024-05-30 at 17 50 00" src="https://github.com/IDinsight/aaq-core/assets/44327443/0bd74853-1e91-4bd6-b760-b351d905f2aa">

### Future Tasks (optional)
- Edit and delete tags
## How has this been tested?
This has been tested by:
- Adding tags to the DB using the backend (Tags need to be added first otherwise the Tags Bar wont be displayed
- Added new contents with tags
- Edited contents by adding/editing tags 
- Tested filtering on landing page
- Tested adding new tag with add "tag"button. 
I suggested testing the feature like this to make sure everything works as intended
## Checklist

Fill with `x` for completed. Delete any lines that are not relevant

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [ ] I have updated the automated tests (if applicable)
- [ ] I have updated the requirements (if applicable)
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)
- [ ] I have added blogpost in Latest Updates

amiraliemami commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -224,6 +240,31 @@ const ContentBox = ({
+            (option) => inputValue === option.tag_name
+          );
+          if (inputValue !== "" && !isExisting) {
+            filtered.push({ tag_id: 0, tag_name: `Add "${inputValue}"` });
```
This is awesome! Can we somehow make sure the tag stays in original form and is not forced to be in caps?

amiraliemami commented on core_backend/app/tags/models.py:
```diff
@@ -18,7 +18,7 @@
 
 content_tags_table = Table(
-    "content_tags",
+    "content-tag",
```
Hey @lickem22, this breaks other relationships - just tried to deploy and getting errors when trying to save content

suzinyou replied:
Instead of rewriting migration script we need a new migration that changes the table name from content_tag to content-tag!

lickem22 replied:
It's not supposed to be an issue because the model still reflects the revision. What erroe do you get? I doubt it is because of the migration script

suzinyou replied:
When I add tags using UI I get this error on the backend:

> sys:1: SAWarning: Object of type <ContentDB> not in session, add operation along 'TagDB.contents' won't proceed (This warning originated from the Session 'autoflush' process, which was invoked automatically in response to a user-initiated operation.)

You are right that I didn't get any errors with the migration step when backend container started.. but why would we change the migration script? I'm not so experienced with DB migrations so just trying to understand. Isn't [core_backend/migrations/versions/2024_05_17_b5ad153a53dc_add_tags.py](https://github.com/IDinsight/aaq-core/pull/221/files#diff-bb14214a7eeaf41c3967d124ed02ddf8bf4d4284341c5043917851b67466e4ce) tied to the latest commit hash on main, and if we want to make any changes to the DB shouldn't we add another migration script?

lickem22 replied:
The goal of the scripts is to have versioning for the database schema just like git. So that when things break, you can revert easily.  Usually when you made a mistake in one of the scripts, you would delete the old migration file and recreate a new migration file. But I decided to update the migration script manually because the result would have been the same whether I generate it or update it manually. 
Also, the error you are getting, I am not able to replicate it

lickem22 replied:
I did some research and basically the reason we have the warning is because the ContentDB model was not added to the session yet. Recreating this warning is a bit tricky but I think I fixed it. Let me know if you still experience it.

amiraliemami commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -224,6 +240,31 @@ const ContentBox = ({
             placeholder="Add Tags"
           />
         )}
+        filterOptions={(options, params) => {
```
I  see that the filtering seems to be an AND, let's do an OR instead as you said? To me showing all contents that have either tag selected in the filter box makes more intutive sense

amiraliemami commented on admin_app/src/app/content/page.tsx:
```diff
@@ -128,13 +179,22 @@ const CardsGrid = ({
-        );
+            card.content_text.toLowerCase().includes(searchTerm.toLowerCase());
+
+          const matchesAllTags = filterTags.every((fTag) =>
+            card.content_tags.includes(fTag.tag_id)
+          );
```
I  see that the filtering seems to be an AND, let's do an OR instead as you said? To me showing all contents that have either tag selected in the filter box makes more intutive sense

lickem22 replied:
I think we agreed on AND as far as I can remember. Let's discuss on Slack

amiraliemami commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -224,6 +240,31 @@ const ContentBox = ({
+        renderOption={(props, option) => {
+          if (option.tag_name) {
+            return (
+              <li {...props}>
+                <Button fullWidth onClick={() => handleNewTag(option.tag_name)}>
+                  {option.tag_name}
+                </Button>
+              </li>
```
What do you think of this? Seems cleaner and gets rid of the weird button-inside-a-list effect

```suggestion
              <li {...props} onClick={() => handleNewTag(option.tag_name)}>
                  {option.tag_name}
              </li>
```

sidravi1 reviewed (changes requested):
Hi @lickem22 - Thanks for this.

I'm having a few issues on my side:

1. I cannot use the mouse to select a tag from the drop down.
2. If i type it in and select it, it throws these warnings in the console:

![image](https://github.com/IDinsight/aaq-core/assets/14125957/71f4553e-9d27-4520-9c6c-a97cd04b6c83)

3. If i create a new tag with the same name as an existing one, it throws an error

![image](https://github.com/IDinsight/aaq-core/assets/14125957/fa3559bc-704d-498a-86e3-efe228b9a9c6)

It should probably handle this cleanly. Maybe (a) don't allow it - remove the add option (b) say already exists

4. When i use the keyboard to select "Add" to create it, it gives something like this:

![image](https://github.com/IDinsight/aaq-core/assets/14125957/18742afc-15de-4a1b-a424-87c8367c7d48)

And doesn't actually create anything in the backend

5. What are your thoughts on creating uppercase and lowercase tags? Right now it allows you to create "hello" and "Hello". I think all tags should be lowercase only and it shouldn't allow the second one and handle it properly (see #3 above). But ok to be overridden if thats what the majority feel is better behaviour.

6. Since the "Save" button is in a different box, it is not obvious to me that you need to hit save. Can we have the state as dirty and throw a warning saying changes will be lost?

7. You probably want to add the props recommended here: https://mui.com/material-ui/react-autocomplete/#creatable

8. Would you also be able to close comment

lickem22 commented:
> * What are your thoughts on creating uppercase and lowercase tags? Right now it allows you to create "hello" and "Hello". I think all tags should be lowercase only and it shouldn't allow the second one and handle it properly (see [Added Tests #3](https://github.com/IDinsight/aaq-core/pull/3) above). But ok to be overridden if thats what the majority feel is better behaviour.
> *

Thanks Sid.
1) Fixed
2) This is a current issue with MUI, I checked online but there is not a fixed for it yet. They are planning to remove it in MUI 6 since it is not a useful warning. (cf this [thread](https://github.com/mui/material-ui/issues/29727#issuecomment-2057764910))
3) Fixed. By checking that the value is not in both availableTags and contentTags
4) Also fixed
5) I think uppercase makes more sense. So now all tags are uppercase. 
6) Not sure I understand that. Are you suggesting we add a warning when leaving the page without saving the updated tags? If we do that for tags, shouldn't we do it for all unsaved updates on contents? Actually I think, if putting the tags outside the contentBox is the issue then we should put it inside. Maybe it is worth discussing with the team and I can update in another PR. 
7) done
8) Not sure I understand that but let me know if I still need to do it

lickem22 commented on admin_app/package-lock.json:
Updated material ui to fix the AutoComplete warning but didn't work. I didn't remove it because i think it does no harm to have the up to date version, but open to reverse the changes.

lickem22 commented on admin_app/src/app/content/edit/page.tsx:
```diff
@@ -224,7 +249,45 @@ const ContentBox = ({
+          );
+        }}
         sx={{ width: "500px" }}
+        isOptionEqualToValue={(option, value) =>
```
It is supposed to fix the tags filter warning. But apparently there is a bug on MUI side. They plan to fix it in the next version

sidravi1 commented:
@lickem22 

1. I think lowercase looks nicer but I'm ok to go with your choice.
2. When i select "Add" using my keyboard (down and enter), it doesn't actually create it but just says "Add "<tag>"". See below

<img width="534" alt="image" src="https://github.com/IDinsight/aaq-core/assets/14125957/38d70a35-4627-46e6-908f-fe922da12990">

leads to 

<img width="562" alt="image" src="https://github.com/IDinsight/aaq-core/assets/14125957/c6bce218-7c84-494d-8e01-0e2ac5e01cc3">

And an error when saving.

3. On #6 above. Yeah! We should do it for all content. Can you raise a ticket to track it? Should be an easy fix since we track it for enabling the "Save" button already.

sidravi1 reviewed (approved):
Thanks @lickem22 - Approving but please merge after you have fixed the issue below.

1. When i select "Add" using my keyboard (down and enter), it doesn't actually create it but just says "Add """. See below

<img width="529" alt="image" src="https://github.com/IDinsight/aaq-core/assets/14125957/abcd8767-78d6-4593-9c3a-66374b057d5d">

leads to

<img width="557" alt="image" src="https://github.com/IDinsight/aaq-core/assets/14125957/b1658783-04f6-4910-bd79-8e8b71702556">

And an error when saving.

2. #6 above. Yeah! We should do it for all content. Can you raise a ticket to track it? Should be an easy fix since we track it for enabling the "Save" button already.

sidravi1 commented on core_backend/tests/api/test_manage_tags.py:
```diff
@@ -70,8 +70,8 @@ async def test_create_and_delete_tag(
         [
-            ("tag_first"),
-            ("tag_second"),
+            ("TAG_FIRST"),
```
We should test if that we cannot create "Tag1" if "tag1" or "TAG1" exist.

sidravi1 replied:
Maybe extend `test_add_tag_same_name_fails`
